proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,334 @@
+slot-height, will be responsible for distributing the rewards.
+
+The leader will pack `n` `DistributeReward` instructions into one transaction.
+`n` will be determined by the maximum size for a transaction divided by the
+size of a `DistributeReward` instruction. In each block, a total of `K/n`
+such transaction will be inserted and propagate down to other nodes in the
+network. All nodes on the network execute these transactions and distribute the
+rewards. In this way, all reward distribution will be recoded and can be traced
+as transactions in the next `M` blocks after reward computation.
+
+One problem is that the stake account can be closed before the reward is paid.
+When this happens, the reward for that account will become unclaimable. To
+prevent this, we propose to add a reward pending flag in the `StakeState` for
+the stake account. When its reward is pending, the stake account wouldn't be
+allowed to close/merge/split. Any of these stake program instructions on the
+stake account with reward pending will fail.
+
+## TODO
+
+The following lists open questions and challenges that either needs a better
+understanding, or needs us to reach a mutually agreed solution.
+
+These items serve as starting points for brainstorming and discussions. As the
+items are resolved, they will be moved to the "design details". Until all the
+items are resolved, the proposal is then merged.
+
+#### Reward calculation Challenges
+
+1. How hard and how expensive is it to clone the stake cache? How significant is
+   the performance overhead for cloning? Are we okay with such overhead?
+
+1. How long does it take to compute the merkle tree? How much time does this
+   add? Is the added time okay for the network?
+
+1. How to adapt the current epoch reward and inflation related RPC methods for
+   new way of computing epoch rewards? What kind of new RPC APIs to provide for
+   to access the `EpochRewardProof`? What's the impact for solana downstream
+   project?
+
+#### Reward distribution challenges
+
+1. What are the incentives for the leader to pack reward redeem transactions?
+   What kind of block fees will the leader receive for packing such
+   transactions? Who will pay for the fees for such transactions? Who will be
+   the signer for those transactions? What prevents a malicious leader from not
+   packing those reward distribution transactions?
+
+   There are two ideas to incentivize the leader to include reward redeeming
+   transactions.
+
+   The first idea is to have the network reject the block that doesn't have all
+   the rewards redeemed. This would requires that all the nodes to keep track of
+   what rewards should be redeemed in a block, and verify that the actual block
+   have the expected rewards. Otherwise, the block will be rejected. The
+   downside of this idea is that it will add extra complexity in block replay
+   stage.
+
+   The second idea is to change up the inflation economics to add a "redeemer
```

Comment:
"redeemer reward" is a big problem because now we need to argue over the economics of the reward, and there'll likely be a lot of people with "input". Best if we can avoid that entirely.

Expecting each block post epoch boundary to distribute N rewards, in a known order, seems way better. I feel this doesn't even need to be done by an instruction in a new program. This can just be bank.rs stuff that happens before, after or concurrently with normal transaction processing for a block

Reply:
Hmm. That sounds like what we have implemented for V1, in which all the rewards are sorted and distributed at the bank "new" time before processing any normal transactions.